modules/rostests/apitests/kernel32/UEFIFirmware.c

```diff
@@ -0,0 +1,221 @@
+        ok(dwLength,
+           "GetFirmwareEnvironmentVariableA failed with error: 0x%08lX\n",
+           GetLastError());
+        if (dwLength)
+        {
+            ok(dwLength == sizeof(EfiVariableValue) && dwValue == EfiVariableValue,
+               "The variable we get is different from what we set, "
+               "dwLength = %lu, dwValue = %lu, expected dwLength = %u, dwValue = %lu",
+               dwLength,
+               dwValue,
+               sizeof(EfiVariableValue),
+               EfiVariableValue);
+        }
+    }
+
+    /* Delete the variable */
+    bResult = SetFirmwareEnvironmentVariableW(_A2W(EFI_TEST_VARIABLE_NAME),
+                                              _A2W(EFI_TEST_GUID_STRING),
+                                              NULL,
+                                              0);
+    ok(bResult,
+       "SetFirmwareEnvironmentVariableW failed with error: 0x%08lX\n",
+       GetLastError());
+    if (bResult)
+    {
+        dwLength = GetFirmwareEnvironmentVariableW(_A2W(EFI_TEST_VARIABLE_NAME),
+                                                   _A2W(EFI_TEST_GUID_STRING),
+                                                   &dwValue,
+                                                   sizeof(dwValue));
+        ok(dwLength == 0, "SetFirmwareEnvironmentVariableW didn't delete the variable!\n");
+    }
+    /* Set variable back and test ANSI version API */
+    bResult = SetFirmwareEnvironmentVariableW(_A2W(EFI_TEST_VARIABLE_NAME),
+                                              _A2W(EFI_TEST_GUID_STRING),
+                                              &EfiVariableValue,
+                                              sizeof(EfiVariableValue));
+    if (!bResult)
+    {
+        skip("SetFirmwareEnvironmentVariableW failed to set variable back with error: 0x%08lX\n",
+             GetLastError());
+        return;
+    }
+    bResult = SetFirmwareEnvironmentVariableA(EFI_TEST_VARIABLE_NAME,
+                                              EFI_TEST_GUID_STRING,
+                                              NULL,
+                                              0);
+    ok(bResult,
+       "SetFirmwareEnvironmentVariableA failed with error: 0x%08lX\n",
+       GetLastError());
+    if (bResult)
+    {
+        dwLength = GetFirmwareEnvironmentVariableA(EFI_TEST_VARIABLE_NAME,
+                                                   EFI_TEST_GUID_STRING,
+                                                   &dwValue,
+                                                   sizeof(dwValue));
+        ok(dwLength == 0, "SetFirmwareEnvironmentVariableA didn't delete the variable!\n");
+    }
+
+#if (_WIN32_WINNT >= 0x0602)
+    /* Test GetFirmwareType, should returns FirmwareTypeUefi if test reached here */
+    FIRMWARE_TYPE FirmwareType = FirmwareTypeUnknown;
+    bResult = GetFirmwareType(&FirmwareType);
+    ok(bResult && FirmwareType == FirmwareTypeUefi,
```

Comment:
Not really: the firmware type can be "Bios" (see MSDN).

Reply:
Committed, both of `FirmwareTypeBios` and `FirmwareTypeUefi` are okay now.